File: frontend/src/types/findings.ts
export type FindingSitemapRefreshNode = {
  url_source_id: number;
  source_refresh_id: number;
  sitemap_document_type: "urlset" | "sitemapindex" | null;
  status: string;
  membership_materialized: boolean;
  children: FindingSitemapRefreshNode[];
};

export type FindingEvaluation = {
  id: number; website_property_id: number; source_scan_id: number | null;
  evaluator_version: string; detector_bundle_identity: string; input_fingerprint_sha256: string;
  evidence_horizon_at: string; active_page_count: number; active_page_universe_sha256: string;
  status: string; detected_count: number; clear_count: number; unknown_count: number;
  detector_summary_json: Record<string, FindingDetectorSummary>;
  evidence_manifest_json: {
    schema?: string;
    static?: { scan_id?: number };
    sitemap_roots?: Array<{
      url_source_id: number;
      refresh_tree: FindingSitemapRefreshNode | null;
    }>;
  };
  created_finding_count: number; resolved_finding_count: number; reopened_finding_count: number;
  assessment_count: number; evaluation_checksum_sha256: string | null; created_at: string;
  started_at: string | null; finished_at: string | null; failed_at: string | null;
  error_type: string | null; error_message: string | null; background_job_id: number | null;
};

export type FindingDetectorSummary = {
  detector_identity: string; detected: number; clear: number; unknown: number;
  reason_counts: Record<string, number>;
};

export type Finding = {
  id: number; web_resource_id: number; page_url: string; finding_type: string;
  finding_label: string;
  logical_key_version: string; fingerprint_sha256: string;
  condition_state: "detected" | "unknown" | "resolved";
  current_severity: "medium" | "high" | null; first_detected_at: string;
  last_detected_at: string; last_evaluated_evidence_at: string; resolved_at: string | null;
  reopened_at: string | null; acknowledged_at: string | null; current_assessment_id: number | null;
  page_workspace_state: string | null; current_evidence_summary: Record<string, unknown>;
};

export type FindingEvidenceReference = {
  id: number; position: number; role: string;
  evidence_kind: "resource_snapshot" | "resource_occurrence" | "source_entry_observation" | "scan"; evidence_id: number;
  evidence_observed_at: string; metadata_json: Record<string, unknown>; retained: boolean; href: string | null;
};

export type FindingAssessment = {
  id: number; finding_evaluation_id: number; outcome: "detected" | "clear" | "unknown";
  severity: "medium" | "high" | null; evidence_observed_at: string;
  details_json: Record<string, unknown>; assessment_sha256: string; created_at: string;
  evaluation: FindingEvaluation; evidence_references: FindingEvidenceReference[];
};

export type FindingDetail = Finding & {
  website_property_id: number; created_at: string; updated_at: string; assessments: FindingAssessment[];
};

export type FindingList = { items: Finding[]; total: number; limit: number; offset: number };
export type FindingEvaluationList = { items: FindingEvaluation[]; total: number; limit: number; offset: number };

export type FindingResetResult = {
  site_id: number;
  deleted_finding_count: number;
  deleted_assessment_count: number;
  deleted_evidence_reference_count: number;
  deleted_evaluation_count: number;
  deleted_job_count: number;
  deleted_job_event_count: number;
};

File: backend/app/schemas/findings.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

FindingState = Literal["detected", "unknown", "resolved"]
FindingSeverity = Literal["medium", "high"]
AssessmentOutcome = Literal["detected", "clear", "unknown"]


class FindingDetectorSummary(BaseModel):
    detector_identity: str
    detected: int
    clear: int
    unknown: int
    reason_counts: dict[str, int] = Field(default_factory=dict)


class FindingEvaluationRead(BaseModel):
    id: int
    website_property_id: int
    source_scan_id: int | None
    evaluator_version: str
    detector_bundle_identity: str
    input_fingerprint_sha256: str
    evidence_horizon_at: datetime
    active_page_count: int
    active_page_universe_sha256: str
    status: str
    detected_count: int
    clear_count: int
    unknown_count: int
    detector_summary_json: dict[str, FindingDetectorSummary] = Field(default_factory=dict)
    created_finding_count: int
    resolved_finding_count: int
    reopened_finding_count: int
    assessment_count: int
    evaluation_checksum_sha256: str | None
    created_at: datetime
    started_at: datetime | None
    finished_at: datetime | None
    failed_at: datetime | None
    error_type: str | None
    error_message: str | None
    background_job_id: int | None = None

    model_config = {"from_attributes": True}


class FindingEvaluationList(BaseModel):
    items: list[FindingEvaluationRead]
    total: int
    limit: int
    offset: int


class FindingListItem(BaseModel):
    id: int
    web_resource_id: int
    page_url: str
    finding_type: str
    finding_label: str
    logical_key_version: str
    fingerprint_sha256: str
    condition_state: FindingState
    current_severity: FindingSeverity | None
    first_detected_at: datetime
    last_detected_at: datetime
    last_evaluated_evidence_at: datetime
    resolved_at: datetime | None
    reopened_at: datetime | None
    acknowledged_at: datetime | None
    current_assessment_id: int | None
    page_workspace_state: str | None
    current_evidence_summary: dict[str, Any] = Field(default_factory=dict)


class FindingList(BaseModel):
    items: list[FindingListItem]
    total: int
    limit: int
    offset: int


class FindingEvidenceReferenceRead(BaseModel):
    id: int
    position: int
    role: str
    evidence_kind: str
    evidence_id: int
    evidence_observed_at: datetime
    metadata_json: dict[str, Any]
    retained: bool
    href: str | None


class FindingAssessmentRead(BaseModel):
    id: int
    finding_evaluation_id: int
    outcome: AssessmentOutcome
    severity: FindingSeverity | None
    evidence_observed_at: datetime
    details_json: dict[str, Any]
    assessment_sha256: str
    created_at: datetime
    evaluation: FindingEvaluationRead
    evidence_references: list[FindingEvidenceReferenceRead]


class FindingDetail(FindingListItem):
    website_property_id: int
    created_at: datetime
    updated_at: datetime
    assessments: list[FindingAssessmentRead]
